### PR TITLE
Use Pybind11 for name functions

### PR DESCRIPTION
**Pull request by sloretz**
Part of #665 

This uses pybind11 for name-related functions. I also added `RCUtilsError` since it was useful for one of the functions, and relaxed test expectations on the `TypeError` human readable text since the text output by pybind11 is a little bit different when an argument is the wrong type.

### CODE REVIEW

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13886/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8739/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11594/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13976/)

**Comment by sloretz** [comment]
Rebased on master

File: rclpy/src/rclpy/names.cpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+  rcutils_allocator_t rcutils_allocator = rcutils_get_default_allocator();
+  rcutils_string_map_t substitutions_map = rcutils_get_zero_initialized_string_map();
+
+  rcutils_ret_t rcutils_ret = rcutils_string_map_init(&substitutions_map, 0, rcutils_allocator);
+  if (rcutils_ret == RCUTILS_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (rcutils_ret != RCUTILS_RET_OK) {
+    throw RCLError("failed to initialize string map");
+  }
```
Comment (cottsay) [inline rclpy/src/rclpy/names.cpp:141]: This should be `rcutils` error handling, not `rcl`.

File: rclpy/src/rclpy/names.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * \return tuple of error message and invalid index if invalid, or
+ * \return None if valid
+ */
+py::object
+get_validation_error_for_namespace(const char * namespace_);
+
+/// Validate a node name and return error message and index of invalidation.
+/**
+ * Raises MemoryError if memory could not be allocated
+ * Raises RuntimeError if an unexpected error happened while validating the node name
```
Comment (cottsay) [inline rclpy/src/rclpy/names.hpp:70]:
```suggestion
 * Raises RCLError if an unexpected error happened while validating the node name
```

File: rclpy/src/rclpy/names.hpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * \param[in] node_name name of the node to be validated
+ * \return tuple of error message and invalid index if invalid, or
+ * \return None if valid
+ */
+py::object
+get_validation_error_for_node_name(const char * node_name);
+
+/// Expand a topic name
+/**
+ * Raises ValueError if the topic name, node name, or namespace are not valid.
```
Comment (cottsay) [inline rclpy/src/rclpy/names.hpp:81]: These last three functions can raise RCLError. Is that omitted intentionally?

**Review by cottsay — CHANGES_REQUESTED** [review]
A few quick comments to address.

It's worth noting that several of the return types are changing from List to Tuple in this PR, though it's probably the right thing to do since it matches the documentation and is more performant.

File: rclpy/src/rclpy/names.cpp (hunk 1 above)
Comment (cottsay) [reply rclpy/src/rclpy/names.cpp:141]: The docs should be updated to reflect the other exception type, too.

File: rclpy/src/rclpy/names.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+  return py::make_tuple(validation_message, invalid_index);
+}
+
+py::object
+get_validation_error_for_namespace(const char * namespace_)
+{
+  int validation_result;
+  size_t invalid_index;
+  rmw_ret_t ret = rmw_validate_namespace(namespace_, &validation_result, &invalid_index);
+  if (ret == RCL_RET_BAD_ALLOC) {
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:91]: @sloretz in the same vein as @cottsay comments about `rcutils` error handling vs. `rcl` error handling, it'd be good to deal with `rmw` error handling using `rmw` API.

File: rclpy/src/rclpy/names.cpp [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  rcl_ret_t ret = rcl_get_default_topic_name_substitutions(&substitutions_map);
+  if (ret == RCL_RET_BAD_ALLOC) {
+    rcl_reset_error();
+    throw std::bad_alloc();
+  } else if (ret != RCL_RET_OK) {
+    auto exception = RCLError("failed to get default substitutions");
+    // finalize the string map before returning
+    rcutils_ret = rcutils_string_map_fini(&substitutions_map);
+    if (rcutils_ret != RCUTILS_RET_OK) {
+      fprintf(
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:152]: @sloretz FYI there's `RCUTILS_SAFE_FWRITE_TO_STDERR` too.

File: rclpy/src/rclpy/names.cpp [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  if (nullptr == node_options) {
+    throw RCLError("failed to get node options");
+  }
+
+  char * output_cstr = NULL;
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:268]: @sloretz nit: 
```suggestion
  char * output_cstr = nullptr;
```

File: rclpy/src/rclpy/names.cpp [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+    };
+  auto resolved_name = std::unique_ptr<char, decltype(name_deleter)>(
+    output_cstr, name_deleter);
+
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to resolve name");
+  }
+
+  return std::string{resolved_name.get()};
+}
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:288]: @sloretz if I'm not mistaken, `output_cstr` will remain a `nullptr` on `rcl_node_resolve_name()` error. In which case I think you can get away with simpler code e.g.:
```suggestion
  if (ret != RCL_RET_OK) {
    throw RCLError("failed to resolve name");
  }

  std::string output{output_cstr};
  node_options->allocator.deallocate(output_cstr, node_options->allocator.state);
  return output;
}
```

**Review by hidmic — COMMENTED** [review]
Overall looks great!

File: rclpy/src/rclpy/names.cpp (hunk 4 above)
Comment (gbalke) [reply rclpy/src/rclpy/names.cpp:91]: I think there a similar issue in `get_validation_error_for_full_topic_name`, `get_validation_error_for_node_name`.

On further investigation, there seems to be some handling for this in rcl. They seem to be translated to rcl error codes though?
https://github.com/ros2/rcl/blob/0ca8c6162136594e490ba69fbf288de1b5f7073d/rcl/src/rcl/expand_topic_name.c#L79

Comment (hidmic) [reply rclpy/src/rclpy/names.cpp:91]: Both `rmw` and `rcl` error handling APIs are aliases of `rcutils`'. That's why people tend to use them interchangeably. Which means we'll have a huge body of code to migrate if that stops being true at some point.

Comment (clalancette) [reply rclpy/src/rclpy/names.cpp:91]: Regardless, I think it is worthwhile to use the correct types (even if they are aliases of each other).  Besides the fact that it makes migration possible (though that is unlikely), there is no question for future readers of the code on whether this is correct or not.

File: rclpy/src/rclpy/names.cpp (hunk 7 above)
Comment (hidmic) [reply rclpy/src/rclpy/names.cpp:288]: Hmm, `std::string` could throw on construction. Nevermind.

File: rclpy/src/rclpy/names.cpp (hunk 5 above)
Comment (hidmic) [reply rclpy/src/rclpy/names.cpp:152]: Oops, I skimmed over the fact you are formatting. Either disregard or drop the integer formatting and use two safe writes.

File: rclpy/src/rclpy/names.cpp (hunk 4 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp]: I added `RMWError` in fc32a0a99ebafc5e28efa343d0e6c868c977d9f7 and used rcl, rcutuils, or rmw constants and functions for each case in c269b6ac92dd3f15aafebffd3117055cfc34cad2

File: rclpy/src/rclpy/names.cpp (hunk 6 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp]: NULL -> nullptr in dcfc7e28d67657fcb329704b7e78e3a88ef38d6c

File: rclpy/src/rclpy/names.hpp (hunk 3 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.hpp:81]: Added note about raising RCLError in c269b6ac92dd3f15aafebffd3117055cfc34cad2

File: rclpy/src/rclpy/names.cpp (hunk 1 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp:143]: Updated docs with RMWError, RCUtilsError, or RCLError in c269b6ac92dd3f15aafebffd3117055cfc34cad2

File: rclpy/src/rclpy/names.hpp (hunk 2 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.hpp]: I think this doc got fixed in c269b6ac92dd3f15aafebffd3117055cfc34cad2 too

File: rclpy/src/rclpy/names.cpp (hunk 5 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp]: Uses two calls to `RCUTILS_SAFE_FWRITE_TO_STDERR` in 90f1052

File: rclpy/src/rclpy/names.cpp (hunk 1 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp]: Oops, actually fixed this one in d7fd3b0d20c4e8602ccdaae09a75b926c5d2d11d

File: rclpy/src/rclpy/names.cpp (hunk 7 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp:292]: Good call on std::string constructor being able to throw. I fixed another spot where I overlooked that in f2d242ab6e3ce2c7cd5230c17bbd99e9f5efec92

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13959/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8792/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11659/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14042/)

File: rclpy/src/rclpy/names.cpp [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    only_expand,
+    &output_cstr);
+
+  auto name_deleter = [&](char * name) {
+      node_options->allocator.deallocate(name, node_options->allocator.state);
+    };
+  auto resolved_name = std::unique_ptr<char, decltype(name_deleter)>(
+    output_cstr, name_deleter);
+
+  if (ret != RCL_RET_OK) {
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:287]: @sloretz nit^N: flip operands, here and elsewhere.

File: rclpy/src/rclpy/names.cpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    node_options->allocator,
+    is_service,
+    only_expand,
+    &output_cstr);
+
+  auto name_deleter = [&](char * name) {
+      node_options->allocator.deallocate(name, node_options->allocator.state);
+    };
+  auto resolved_name = std::unique_ptr<char, decltype(name_deleter)>(
+    output_cstr, name_deleter);
```
Comment (hidmic) [inline rclpy/src/rclpy/names.cpp:285]: @sloretz FYI, nit^N: [`RCPPUTILS_SCOPE_EXIT`](https://github.com/ros2/rcpputils/blob/71ceb6b0597eee71f91dfeffe77fca7cc1a9cdb4/include/rcpputils/scope_exit.hpp#L65) may be handy in these cases.

**Review by hidmic — APPROVED** [review]
LGTM but for a few minor nits I won't block on.

File: rclpy/src/rclpy/names.cpp (hunk 8 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp]: Flipped conditionals in 193a026

**Comment by sloretz** [comment]
CI (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13961/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8794/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11661/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14044/)

File: rclpy/src/rclpy/names.cpp (hunk 9 above)
Comment (sloretz) [reply rclpy/src/rclpy/names.cpp:285]: Noted, I'll use this on the next PR once the rcpputils dependency is added.

**Review by hidmic — APPROVED** [review]

**Pull request merged**